src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs

```diff
@@ -635,73 +637,78 @@ private CancellationTokenRegistration RegisterCancellation(CancellationToken can
-                line[2] != 'T' ||
-                line[3] != 'P' ||
-                line[4] != '/' ||
-                line[5] != '1' ||
-                line[6] != '.' ||
-                line[8] != ' ')
+            // Validate the beginning of the status line and set the response version.
+            const int MinStatusLineLength = 12; // "HTTP/1.x 123" 
+            if (line.Length < MinStatusLineLength || line[8] != ' ')
             {
                 ThrowInvalidHttpResponse();
             }
 
-            // Set the response HttpVersion
-            byte minorVersion = line[7];
-            response.Version =
-                minorVersion == '1' ? HttpVersion.Version11 :
-                minorVersion == '0' ? HttpVersion.Version10 :
-                !IsDigit(minorVersion) ? throw new HttpRequestException(SR.net_http_invalid_response) :
-                new Version(1, minorVersion - '0');
+            ulong first8Bytes = BitConverter.ToUInt64(line);
+            if (first8Bytes == s_http11Bytes)
+            {
+                response.SetVersionUnchecked(HttpVersion.Version11);
+            }
+            else if (first8Bytes == s_http10Bytes)
+            {
+                response.SetVersionUnchecked(HttpVersion.Version10);
+            }
+            else
+            {
+                byte minorVersion = line[7];
+                line[7] = (byte)'1'; // Enable checking the first 7 bytes easily by comparing against HTTP/1.1
+                if (IsDigit(minorVersion) && BitConverter.ToUInt64(line) == s_http11Bytes)
+                {
+                    response.SetVersionUnchecked(new Version(1, minorVersion - '0'));
+                }
+                else
+                {
+                    ThrowInvalidHttpResponse();
+                }
+            }
 
             // Set the status code
             byte status1 = line[9], status2 = line[10], status3 = line[11];
             if (!IsDigit(status1) || !IsDigit(status2) || !IsDigit(status3))
             {
                 ThrowInvalidHttpResponse();
             }
-
-            response.StatusCode =
-                (HttpStatusCode)(100 * (status1 - '0') + 10 * (status2 - '0') + (status3 - '0'));
+            response.SetStatusCodeUnchecked((HttpStatusCode)(100 * (status1 - '0') + 10 * (status2 - '0') + (status3 - '0')));
 
             // Parse (optional) reason phrase
             if (line.Length == MinStatusLineLength)
             {
                 response.ReasonPhrase = string.Empty;
             }
-            else if (line[MinStatusLineLength] != ' ')
+            else if (line[MinStatusLineLength] == ' ')
             {
-                ThrowInvalidHttpResponse();
+                Span<byte> reasonBytes = line.Slice(MinStatusLineLength + 1);
+                string knownReasonPhrase = HttpStatusDescription.Get(response.StatusCode);
+                response.SetReasonPhraseUnchecked(knownReasonPhrase != null && EqualsOrdinal(knownReasonPhrase, reasonBytes) ?
```

Comment:
Seems like we should still validate unknown reason phrases -- either add the logic here specifically and throw on failure, or just call the checked API.

See #26946 also, which is probably a good argument for adding the logic here.

Reply:
> Seems like we should still validate unknown reason phrases

We already did.  The only thing it validates is that it doesn't contain \r or \n and that it's not null, and none of those could be the case here.

Reply:
Can we reenable the test for #26946 and close the issue, then?

Reply:
Actually, I'm wrong.  I thought our line parsing was looking for \r, but it's now looking for \n, which means a \r could still be here.  I'll make the unknown phrase still use the checking logic.

Reply:
(but I'll fix the issue with that test and re-enable it anyway)